module/Database/src/Service/Payment.php

```diff
@@ -0,0 +1,313 @@
+        if (
+            CheckoutSessionModel::PAID === $lastCheckoutStub->getState()
+            || CheckoutSessionModel::PENDING === $lastCheckoutStub->getState()
+        ) {
+            // Checkout Session is finalised or will be after payment processing. Do not allow the prospective member to
+            // do something else.
+            return null;
+        }
+
+        if (CheckoutSessionModel::FAILED === $lastCheckoutStub->getState()) {
+            // Last payment failed, so we need to create a new Checkout Session for the user to be able to try again.
+            return $this->getCheckoutLink($prospectiveMember);
+        }
+
+        if (CheckoutSessionModel::EXPIRED === $lastCheckoutStub->getState()) {
+            // The Checkout Session has already been abandoned.
+
+            if ((new DateTime())->add(new DateInterval('PT5M')) >= $lastCheckoutStub->getExpiration()) {
+                // The Checkout Session is completely abandoned, as the maximum expiration for the recovery URL of 30
+                // days has passed (or will pass in 5 minutes). As such, we want to create a new Checkout Session for
+                // this prospective member.
+                return $this->getCheckoutLink($prospectiveMember);
+            }
+
+            // The Checkout Session is not completely dead yet, so return the recovery URL.
+            return $lastCheckoutStub->getRecoveryUrl();
+        }
+
+        // Checkout Session is still valid (at this point, not necessarily when the prospective member finally submits).
+        // We try to retrieve the actual Checkout Session, if this succeeds we return the URL. If not we fail and let
+        // the prospective member know.
+        return $this->getCheckoutSession($lastCheckoutStub->getCheckoutId())?->url;
+    }
+
+    /**
+     * Create a Checkout Session through the Stripe API. The Checkout Session has some required parameters, for more
+     * information on their details see {@link https://stripe.com/docs/api/checkout/sessions/object}.
+     */
+    private function createCheckoutSession(ProspectiveMemberModel $prospectiveMember): ?CheckoutSession
+    {
+        try {
+            return $this->getClient()->checkout->sessions->create([
+                'line_items' => [
+                    [
+                        'price' => $this->config['membership_price_id'],
+                        'quantity' => 1,
+                    ],
+                ],
+                'mode' => 'payment',
+                'cancel_url' => $this->config['cancel_url'],
+                'success_url' => $this->config['success_url'],
+                'after_expiration' => [
+                    'recovery' => [
+                        'enabled' => true,
+                    ],
+                ],
+                'client_reference_id' => $prospectiveMember->getLidnr(),
+                'customer_email' => $prospectiveMember->getEmail(),
```

Comment:
I do not think we want to always create a `customer`, so `customer_creation` can stay `if_required`. Stripe will likely only create `Guest` customers for card payments. Adding more information about the payer, such as full name, is therefore not needed as that requires the creation of a `customer`.

Reply:
I agree